project/src/main/webapp/gmail.js

```diff
@@ -0,0 +1,242 @@
+    }
+
+    /**
+      * Append a pre element to the body containing the given message
+      * as its text node. Used to display the results of the API call.
+      *
+      * @param {string} message Text to be placed in pre element.
+      */
+    function appendPre(message) {
+        var pre = document.getElementById('content');
+        var textContent = document.createTextNode(message + '\n');
+        pre.appendChild(textContent);
+    }
+
+    /**
+      * Print all Labels in the authorized user's inbox. If no labels
+      * are found an appropriate message is printed.
+      */
+    // function listLabels() {
+    //     const t0 = performance.now();
+    //     gapi.client.gmail.users.labels.list({
+    //       'userId': 'me'
+    //     }).then(function(response) {
+    //         var labels = response.result.labels;
+    //         appendPre('Labels:');
+
+    //         if (labels && labels.length > 0) {
+    //             for (i = 0; i < labels.length; i++) {
+    //             var label = labels[i];
+    //             appendPre(label.name);
+    //             console.log(label.name);
+    //             }
+    //         } else {
+    //             appendPre('No Labels found.');
+    //         }
+    //     const t1 = performance.now();
+    //     console.log(t1-t0);
+    //     });
+    // }
+
+    // function listMessages() {
+    //     const t0 = performance.now();
+    //     gapi.client.gmail.users.messages.list({
+    //       'userId': 'me'
+    //     }).then(function(response) {
+    //         var messages = response.result.messages;
+    //         //appendPre('Messages:');
+
+    //         if (messages && messages.length > 0) {
+    //             for (i = 0; i < messages.length; i++) {
+    //             var message = messages[i];
+    //             //appendPre(message.name);
+    //             console.log(message.name);
+    //             }
+    //         } else {
+    //             appendPre('No Messages found.');
+    //         }
+    //     const t1 = performance.now();
+    //     console.log(t1-t0);
+    //     });
+    // }
+
+    async function keywordExtraction(userInput) {
+    var result = "";
+    var count = 0;
+    var searchQuery = userInput.split(" ");
+    for (word of searchQuery){
+        if ((searchQuery.length-1) == count && word.indexOf('?') != -1) {
+                word = word.substring(0, word.length-1);
+            }
+        var url = new URL("https://api.datamuse.com/words?md=p&sp=" + word + "&max=1");
+        const output = await fetch(url).then(response => response.text());
+        if (output.includes("[\"n\"]") == true || output.includes("[\"adj\"]") == true){
+            result += word;
+            if (count != (searchQuery.length-1)){
+                result += " ";
+            }
+        }
+        count++;
+    }
+    return result; 
+    }
+    
+    //Retrieve messages using hardcoded queries and the signed-in email.
+    async function listMessages() {
```

Comment:
Break up function to get rid of nested statements(loops,if statements, promises). Line 202 could be a separate function(if I am reading it correctly).